activesupport/lib/active_support/testing/instrumentation_assertions.rb

```diff
@@ -0,0 +1,80 @@
+
+# typed: true
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Testing
+    module InstrumentationAssertions
+      # Assert an event was emitted with a given +pattern+ and optional +payload+.
+      #
+      # You can assert that an event was emitted by passing a pattern, which accepts
+      # either a string or regexp, an optional payload, and a block. While the block
+      # is executed, if a matching event is emitted, the assertion will pass.
+      #
+      #     assert_instrumentation_event("post.submitted", title: "Cool Post") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_instrumentation_event(pattern, payload = nil, &block)
+        events = capture_instrumentation_events(pattern, &block)
+        assert_not_empty(events, "No #{pattern} events were found.")
+
+        return if payload.nil?
+
+        event = events.find { |event| event.payload == payload }
+        assert_not_nil(
+          event,
+          "No #{pattern} event with payload #{payload} was found. " \
+            "Did you mean one of these payloads: #{events.map(&:payload).join("\n")}?",
```

Comment:
Should there be a line return before the first suggested payload?
```suggestion
            "Did you mean one of these payloads:\n#{events.map(&:payload).join("\n")}?",
```